### PR TITLE
ENH: use neurodocker 0.4.1 + apt install afni

### PR DESCRIPTION
**Pull request by kaczmarj**
## Summary

This PR proposes updates to Dockerfile generation.

References https://github.com/nipy/nipype/pull/2706#issuecomment-422875107

## List of changes proposed in this PR (pull-request)
- upgrade to neurodocker 0.4.1
- install afni from neurodebian
- minor refactoring in `neurodocker generate` commands.

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by effigies** [comment]
This seems to be broken by #2691...

**Comment by effigies** [comment]
@kaczmarj Is there a trivial way to trigger a new docker build? I want to make sure that things aren't only passing on master because we're reusing the old images.

**Comment by effigies** [comment]
Built from scratch on `rel/1.1.3` locally and no issues. Will try your branch now.

**Comment by effigies** [comment]
Yeah, that fails. Looks like you're changing the Matlab and/or SPM installs, and our doc parsing is failing for these versions. How about we push this off to 1.1.4? Or possibly restore the old Matlab/SPM.

**Comment by kaczmarj** [comment]
@effigies - i think the issue is with the matlab compiler runtime installation. i'm fine pushing this to the next release, but i'll fix it soon.

**Comment by effigies** [comment]
Want to merge master? Circle seems to have done okay on Oscar's coverage update.

**Comment by kaczmarj** [comment]
The following command is failing.

```
flameo \
--copefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/copemerge/mapflow/_copemerge0/cope1_merged.nii.gz \
--covsplitfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.grp \
--designfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.mat \
--dofvarcopefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/gendofvolume/dof_file.nii.gz \
--ld=stats \
--maskfile=/work/output/fmri_fsl_reuse/Linear/level1flow/featpreproc/_subject_id_s1/_fwhm_5.0/dilatemask/mapflow/_dilatemask0/f3_dtype_mcf_bet_thresh_dil.nii.gz \
--runmode=fe \
--tcontrastsfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.con \
--varcopefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/varcopemerge/mapflow/_varcopemerge0/varcope1_merged.nii.gz
```

One can run the following to reproduce
```
/usr/bin/run_examples.sh fmri_fsl_reuse Linear /data/examples/ level1_workflow
```

**Comment by kaczmarj** [comment]
This is the specific error:

```
flameo --copefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/copemerge/mapflow/_copemerge0/cope1_merged.nii.gz --covsplitfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.grp --designfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.mat --dofvarcopefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/gendofvolume/dof_file.nii.gz --ld=stats --maskfile=/work/output/fmri_fsl_reuse/Linear/level1flow/featpreproc/_subject_id_s1/_fwhm_5.0/dilatemask/mapflow/_dilatemask0/f3_dtype_mcf_bet_thresh_dil.nii.gz --runmode=fe --tcontrastsfile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/l2model/design.con --varcopefile=/work/output/fmri_fsl_reuse/Linear/level1flow/fixedfx/_subject_id_s1/_fwhm_5.0/varcopemerge/mapflow/_varcopemerge0/varcope1_merged.nii.gz
Standard output:
Log directory is: stats
Setting up:
ntptsing=4.000000 

evs_group=1.000000 

Standard error:
terminate called after throwing an instance of 'RBD_COMMON::BaseException'
Aborted (core dumped)
Return code: 134
```

**Comment by effigies** [comment]
Neurodebian's FSL doesn't look to have been updated (correct @yarikoptic?). I wonder if there's a library bug that the new builds are hitting.

**Comment by effigies** [comment]
Looks like the build is fixed on master. Unclear why.

**Comment by effigies** [comment]
Ah, I see. Master is able to reuse the old base images, because of the compare_dockerfiles check. PRs from most users won't be able to do so. So fresh builds are still failing, we're just not doing them on master.

**Comment by effigies** [comment]
I reproduced the issue. Adding `--verbose --debug=2` gives practically no more output. I see a couple options:

1) Try non-Neurodebian FSL to verify that the issue persists. (Possibly upgrade versions?)
2) Ping the FSL mailing list.

Any other thoughts?

**Comment by kaczmarj** [comment]
@effigies - i will see if using a neurodebian snapshot from a few months ago will solve the issue http://snapshot.debian.org/

**Comment by effigies** [comment]
@kaczmarj Any luck here? Is there something I can do to help this along?

**Comment by effigies** [comment]
@kaczmarj Any chance we can figure this one out this week? Tests are still failing for all new PRs, and it would be nice to have this fixed before the next release.

**Comment by kaczmarj** [comment]
@effigies - i'm working on this now

**Comment by effigies** [comment]
Thanks. Sorry to pester.

**Comment by effigies** [comment]
Did some of my own tests on #2763. Even installing directly from FSL, flameo crashes for both 5.0.9 and 5.0.11. Perhaps it's a libc issue?

IDK. Is there any chance of just fixing the broken aspects of the current Dockerfile? Or is this a minimal update already?

**Comment by kaczmarj** [comment]
@effigies - this is a minimal update already. i tried to update the original (neurodocker 0.3.2) dockerfile to use the non-neurodebian fsl to no avail.

File: docker/generate_dockerfiles.sh [hunk 1]
```diff
@@ -52,77 +52,70 @@ do
...
-
-# neurodocker version 0.3.1-22-gb0ee069
-NEURODOCKER_IMAGE="kaczmarj/neurodocker@sha256:f15ca90803f4b89acfca55cd1c7269bf2ec2dfd95b3de1118b08afa34b87d9d1"
-
-# neurodebian:stretch-non-free pulled on November 3, 2017
-BASE_IMAGE="neurodebian@sha256:7590552afd0e7a481a33314724ae27f76ccedd05ffd7ac06ec38638872427b9b"
+# neurodocker version 0.4.1-22-g7c44e01
+NEURODOCKER_IMAGE="kaczmarj/neurodocker:master@sha256:858632a7533cac100f70932749b4cfc77fc40f667f41fca208f406215cff8a27"
+# neurodebian:stretch-non-free pulled on September 19, 2018
+BASE_IMAGE="neurodebian:stretch-non-free@sha256:7cd978427d7ad215834fee221d0536ed7825b3cddebc481eba2d792dfc2f7332"
```
Comment (effigies) [inline docker/generate_dockerfiles.sh]: @kaczmarj What if we revert the `BASE_IMAGE`? If that helps, I think that would be decent evidence that it's a library version issue.

Comment (kaczmarj) [reply docker/generate_dockerfiles.sh]: @effigies - i don't think that will help. the `apt-get update` calls will update to the latest neurodebian stretch sources. i tried using neurodebian snapshots from september 17, 2017 and from january 1, 2017, but both ran into the same `flameo` error. i'm going to try to use the nipype base image on dockerhub with the current nipype source code now.

Comment (kaczmarj) [reply docker/generate_dockerfiles.sh]: @effigies - i used nipype's base image on dockerhub (i.e., `nipype/nipype:base`) and still got the `flameo` error. is it possible that nipype's code is causing this? could one of the inputs to `flameo` be corrupted / malformed?

Comment (effigies) [reply docker/generate_dockerfiles.sh]: But it works on `master`, still, with the same inputs. There's something new in the build.

Comment (effigies) [reply docker/generate_dockerfiles.sh]: @kaczmarj Is there any chance of using a Debian snapshot from earlier, as well? If it's a libc update or similar that's causing the issue, that wouldn't be resolved by using a neurodebian snapshot.

Comment (kaczmarj) [reply docker/generate_dockerfiles.sh]: true, i'll try that now. i'm experimenting with updating packages in the current nipype py36 docker image and see if that will reproduce the error.

**Comment by kaczmarj** [comment]
@effigies - i've narrowed down the issue to the SPM12 matlab r2010a installation in neurodocker 0.4.1. unfortunately i don't have time to dig into it right this minute, but i will this evening / tomorrow morning.

**Comment by miykael** [comment]
@kaczmarj, I remember that I had some library issues when I switched the tutorial from spm-dev to spm-r7219. I think you showed me this, but the solution in my case was to first look through the `miniconda` libraries before going into the ones from MATLAB, see [here](https://github.com/miykael/nipype_tutorial/blob/master/generate.sh#L24).

**Comment by kaczmarj** [comment]
thanks @miykael - this indeed was the issue. the most recent circleci build should pass now (it did on my computer).

**Comment by effigies** [comment]
Sweet! Here we go.

**Comment by kaczmarj** [comment]
Wahoo!

And as an added bonus, it takes about 9 minutes to build both the base docker image and the main (py36) image. according to @miykael, the matlab compiler runtime file for matlab 2010a is much smaller than that for 2018b, so using 2010a decreases the spm/mcr installation significantly. that and afni being installed from neurodebian.

**Comment by miykael** [comment]
Exactly. Switching to SPM-r7219 decreases MATLAB by a few GB and even SPM by a fe 100MB. And afni (and ants) from neuredebian has a similar reduction effect. Like this I was able to get the tutorial from 12GB to 6GB.

**Pull request merged**